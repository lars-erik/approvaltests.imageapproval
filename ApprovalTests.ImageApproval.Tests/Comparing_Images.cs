using System.IO;
using ApprovalTests;
using ApprovalTests.Core.Exceptions;
using ApprovalTests.ImageApproval;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using ApprovalTests.Reporters.TestFrameworks;
using ApprovalTests.Writers;
using Newtonsoft.Json;
using NUnit.Framework;
using SixLabors.ImageSharp.Formats.Png;

[assembly: UseApprovalSubdirectory("Approvals")]

namespace ApprovalTestContrib.ImageApproval.Tests
{
    [UseReporter(typeof(NUnitAttachmentReporter), typeof(NUnitImageComparisonReporter))]
    [TestFixture]
    public class Comparing_Images
    {
        [Test]
        public void Approves_Equal_Images()
        {
            var imageBytes = GetComparisonImage("ApprovalTestsRocks");
            ImageApprovals.Verify(imageBytes, "png");
        }

        [Test]
        public void Approves_Image_With_Tolerance()
        {
            var imageBytes = GetComparisonImage("ApprovalTestsRocks.Nudged");
            ImageApprovals.Verify(imageBytes, "png", 0.021);
        }

        [Test]
        public void Disapproves_Unequal_Images()
        {
            var imageBytes = GetComparisonImage("ApprovalTestsRocks.Nudged");
            Assert.That(() => ImageApprovals.Verify(imageBytes, "png"), Throws.TypeOf<AssertionException>());
        }

        [Test]
        public void Stores_Color_Diff_In_State()
        {
            var imageBytes = GetComparisonImage("ApprovalTestsRocks.Nudged");
            ImageApprovals.Verify(imageBytes, "png", 0.021);

            using (ApprovalResults.ForScenario("diff"))
            {
                using (var str = new MemoryStream())
                {
                    ImageApprovals.Current.ColorDiff.Save(str, new PngEncoder());
                    str.Seek(0, SeekOrigin.Begin);
                    var bytes = new byte[str.Length];
                    str.Read(bytes, 0, (int)str.Length);
                    Approvals.VerifyBinaryFile(bytes, "png");
                }
            }
        }

        [Test]
        [UseReporter(typeof(NUnitReporter))]
        public void Stores_Json_Diff_In_State()
        {
            var imageBytes = GetComparisonImage("ApprovalTestsRocks.Nudged");
            ImageApprovals.Verify(imageBytes, "png", 0.021);

            Approvals.VerifyJson(JsonConvert.SerializeObject(ImageApprovals.Current.Statistics));
        }

        private byte[] GetComparisonImage(string fileName)
        {
            byte[] result;
            using (var stream = GetType().Assembly
                .GetManifestResourceStream(
                    $"ApprovalTestContrib.ImageApproval.Tests.Comparison_Images.{fileName}.png"))
            {
                result = new byte[stream.Length]; // TODO: Handle too big streams
                stream.Read(result, 0, (int)stream.Length);
            }

            return result;
        }
    }
}