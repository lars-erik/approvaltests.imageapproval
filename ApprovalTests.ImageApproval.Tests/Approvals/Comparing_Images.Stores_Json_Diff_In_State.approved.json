{
    "CountedByNoticable": {
        "False": 2032558,
        "True": 41042
    },
    "AveragedByNoticable": {
        "True": 2.6603648819698593,
        "False": 3.1249469523024733E-05
    },
    "SummedByNoticable": {
        "True": 354272.81023728027,
        "False": 60.63750189497146
    },
    "CountedByCie76Floored": {
        "0": 2021310,
        "1": 10231,
        "7": 6964,
        "8": 27932,
        "4": 1123,
        "3": 1218,
        "6": 1306,
        "5": 1075,
        "2": 2441
    },
    "AveragedByCie76Floored": {
        "0": 0.011873622913369094,
        "1": 1.373998011751022,
        "7": 7.699452982623563,
        "8": 8.346170976156587,
        "4": 4.487720822210555,
        "3": 3.4549349532588374,
        "6": 6.5441112424502785,
        "5": 5.464983027903528,
        "2": 2.401587783236058
    },
    "SummedByCie76Floored": {
        "0": 24000.272731022083,
        "1": 14057.373658224706,
        "7": 53618.99057099049,
        "8": 233125.2477060058,
        "4": 5039.710483342454,
        "3": 4208.110773069264,
        "6": 8546.609282640064,
        "5": 5874.856754996292,
        "2": 5862.275778879218
    },
    "MaxCie76": 8.65023731448525,
    "AvgCie76": 0.17087839879397185,
    "SumCie76": 354333.44773918,
    "NoticableRatio": 0.02019228971571783,
    "ToleratedRato": 0.021,
    "ExceedsRatio": false
}